File: minebridge/minebridge-core/src/main/java/org/mineacademy/minebridge/core/schema/SendPlayerMessage.java
package org.mineacademy.minebridge.core.schema;

import org.mineacademy.minebridge.core.internal.BaseSchema;

import lombok.Getter;

public class SendPlayerMessage extends BaseSchema {

    /**
     * The username of the player to send the message to.
     */
    @Getter
    private final String username;

    /**
     * The UUID of the player to send the message to.
     */
    @Getter
    private final String uuid;

    /**
     * The message type to send to the player.
     */
    @Getter
    private final String message_type;

    /**
     * The message to send to the player.
     */
    @Getter
    private final String message;

    public SendPlayerMessage(String username, String uuid, String message_type, String message) {
        super("send-player-message", false);
        this.username = username;
        this.uuid = uuid;
        this.message_type = message_type;
        this.message = message;
    }

}

File: minebot/src/websocket/schemas/request/send_player_message.py
from pydantic import Field, model_validator

from model import MessageType

from ..base import BaseSchema, ResponseAwaitableSchema


class SendPlayerMessageSchema(BaseSchema, ResponseAwaitableSchema, action="send-player-message"):
    """Schema for the send-player-message action."""

    message_type: MessageType
    message: str
    username: str | None = Field(default=None, max_length=16)
    uuid: str | None = Field(default=None, max_length=36)

    @model_validator(mode="after")
    def validate_username_or_uuid(self) -> "SendPlayerMessageSchema":
        if self.username is None and self.uuid is None:
            raise ValueError("Either 'username' or 'uuid' must be provided.")
        return self
